File: Sources/NotchAgent/Core/Models/Events.swift
import Foundation

public enum AttentionLevel: Int, Codable, Sendable, Comparable, CaseIterable {
    case normal = 0
    case warning = 1
    case critical = 2

    public static func < (lhs: AttentionLevel, rhs: AttentionLevel) -> Bool {
        lhs.rawValue < rhs.rawValue
    }

    public var label: String {
        switch self {
        case .normal: "Normal"
        case .warning: "Warning"
        case .critical: "Critical"
        }
    }
}

public struct ProviderAlert: Sendable, Equatable {
    public var provider: ProviderID
    public var level: AttentionLevel
    public var message: String
    public var date: Date

    public init(provider: ProviderID, level: AttentionLevel, message: String, date: Date = Date()) {
        self.provider = provider
        self.level = level
        self.message = message
        self.date = date
    }
}

public struct UsageEvent: Codable, Sendable, Identifiable, Equatable {
    public enum Kind: String, Codable, Sendable, CaseIterable {
        case alert
        case error
        case info
    }

    public var id: UUID
    public var date: Date
    public var provider: ProviderID?
    public var kind: Kind
    public var level: AttentionLevel
    public var message: String

    public init(
        id: UUID = UUID(),
        date: Date = Date(),
        provider: ProviderID? = nil,
        kind: Kind,
        level: AttentionLevel = .normal,
        message: String
    ) {
        self.id = id
        self.date = date
        self.provider = provider
        self.kind = kind
        self.level = level
        self.message = message
    }
}

public enum RefreshState: Sendable, Equatable {
    case idle
    case refreshing
    case success(Date)
    case failure(Date, String)
}

File: windows/NotchAgent.Windows/Models/Events.cs
namespace NotchAgent.Windows.Models;

public sealed class UsageEvent
{
    public enum EventKind { Alert, Error, Info }

    public Guid Id { get; } = Guid.NewGuid();
    public DateTimeOffset Date { get; init; } = DateTimeOffset.UtcNow;
    public ProviderId? Provider { get; init; }
    public required EventKind Kind { get; init; }
    public AttentionLevel Level { get; init; } = AttentionLevel.Normal;
    public required string Message { get; init; }
}

public sealed record ProviderAlert(ProviderId Provider, AttentionLevel Level, string Message)
{
    public DateTimeOffset Date { get; } = DateTimeOffset.UtcNow;
}

/// Escalating "space left" alert fired when a provider's gauge crosses below
/// 25 / 15 / 10 / 5 percent remaining. Each threshold fires once per window;
/// everything re-arms when the window resets (remaining climbs back up).
public sealed record ThresholdAlert(ProviderId Provider, int Threshold, double Remaining, bool IsWeekly)
{
    public DateTimeOffset FiredAt { get; } = DateTimeOffset.UtcNow;
}

/// The positive counterpart to a low-fuel alert: fired once when a provider
/// transitions from blocked back to usable while the user was recently active.
public sealed record RestoreMoment(ProviderId Provider, double Remaining, bool IsWeekly)
{
    public DateTimeOffset FiredAt { get; } = DateTimeOffset.UtcNow;

    public string Message =>
        $"{Provider.DisplayName()} is back — {Math.Round(Remaining)}% of the {(IsWeekly ? "weekly limit" : "5h session")} left";
}
